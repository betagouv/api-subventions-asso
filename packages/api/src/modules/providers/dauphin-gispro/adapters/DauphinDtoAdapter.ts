import { ApplicationNature, ApplicationStatus, DocumentDto, ProviderValue } from "dto";
import ProviderValueFactory from "../../../../shared/ProviderValueFactory";
import dauphinService from "../dauphin.service";
import { toStatusFactory } from "../../providers.adapter";
import DauphinDocumentDto from "../dto/DauphinDocumentDto";
import { ApplicationFlatEntity } from "../../../../entities/ApplicationFlatEntity";
import { SimplifiedJoinedDauphinGispro } from "../@types/SimplifiedDauphinGispro";
import { GenericAdapter } from "../../../../shared/GenericAdapter";
import Siret from "../../../../identifierObjects/Siret";

export class InconsistentAggregationError extends Error {
    public field: string;
    public valueList: unknown[];
    public codeDossier: string | undefined;
    public referenceAdministrative: string[];

    constructor(
        field: string,
        valueList: unknown[],
        codeDossier: string | undefined,
        referenceAdministrative: string[],
    ) {
        const message = `Les valeurs suivantes du champ ${field} devraient être identiques mais ne le sont pas pour le dossier de code ${codeDossier} et les codes action ${referenceAdministrative} : ${valueList}`;
        super(message);
        this.field = field;
        this.valueList = valueList;
        this.codeDossier = codeDossier;
        this.referenceAdministrative = referenceAdministrative;
    }
}

export default class DauphinDtoAdapter {
    private static _statusConversionArray: { label: ApplicationStatus; providerStatusList: string[] }[] = [
        { label: ApplicationStatus.REFUSED, providerStatusList: ["Rejetée"] },
        {
            label: ApplicationStatus.GRANTED,
            providerStatusList: [
                "A justifier",
                "Justifiée",
                "A été versé",
                "Justification à modifier",
                "Justification en cours",
            ],
        },
        { label: ApplicationStatus.INELIGIBLE, providerStatusList: ["Cloturée", "Non recevable"] },
        {
            label: ApplicationStatus.PENDING,
            providerStatusList: [
                "Prise en charge",
                "Recevable",
                "Transmise",
                "En attente d'attestation",
                "En attente d'instruction",
                "En cours de saisie",
                "En cours",
            ],
        },
    ];

    private static getStatus(rawStatus: string) {
        return toStatusFactory(DauphinDtoAdapter._statusConversionArray)(rawStatus);
    }

    public static toDocuments(documentReferences: DauphinDocumentDto[]) {
        const resultArray: DocumentDto[] = [];
        let toPV: <T>(T) => ProviderValue<T>;
        for (const reference of documentReferences) {
            if (!reference.documents?.length) continue;
            for (const doc of reference.documents) {
                if (doc.error) continue;
                const date = new Date(doc.expand.properties["entity:document:date"].value);
                toPV = ProviderValueFactory.buildProviderValueAdapter(dauphinService.meta.name, new Date(date));
                resultArray.push({
                    type: toPV(reference.libelle.value),
                    url: toPV(
                        `/document/dauphin/?url=${encodeURIComponent(`https://agent-dauphin.cget.gouv.fr${doc.id}`)}`,
                    ),
                    nom: toPV(doc.title || reference.libelle.value),
                    __meta__: {},
                });
            }
        }
        return resultArray;
    }

    public static simplifiedJoinedToApplicationFlat(simplified: SimplifiedJoinedDauphinGispro): ApplicationFlatEntity {
        const getSingleValueOrNull = <T>(valueList: T[]) => (valueList.length === 1 ? valueList[0] : null);
        const getSingleValueOrThrow = <T>(valueList: T[], field: string) => {
            if (valueList.length === 0) return null;
            if (valueList.length === 1) return valueList[0];
            throw new InconsistentAggregationError(
                field,
                valueList,
                simplified.codeDossier,
                simplified.referenceAdministrative,
            );
        };

        const localId = simplified.referenceAdministrative.join("|");
        const dateDemandeStr = getSingleValueOrNull(simplified.dateDemande);
        const dateDemande = dateDemandeStr ? new Date(dateDemandeStr) : null;
        const status = getSingleValueOrThrow(
            [...new Set(simplified.virtualStatusLabel.map(this.getStatus))],
            "virtualStatusLabel",
        ) as ApplicationStatus;
        const ej = getSingleValueOrNull(simplified.ej);

        const adapted = {
            allocatorId: null,
            allocatorIdType: null,
            allocatorName: simplified.financeurs.join("|"),
            applicationId: "dauphin-" + localId,
            applicationProviderId: localId,
            beneficiaryEstablishmentId: simplified.siretDemandeur,
            beneficiaryEstablishmentIdType: Siret.getName(),
            budgetaryYear: simplified.exerciceBudgetaire,
            cofinancersIdType: null,
            cofinancersNames: null,
            cofinancingRequested: null,
            confinancersId: null,
            conventionDate: null,
            decisionDate: null,
            decisionReference: null,
            depositDate: dateDemande,
            ej,
            idRAE: null,
            instructiveDepartementId: null,
            instructiveDepartmentIdType: null,
            instructiveDepartmentName: simplified.instructorService.join("|"),
            joinKeyDesc: null,
            joinKeyId: null,
            managingAuthorityId: GenericAdapter.NOT_APPLICABLE_VALUE,
            managingAuthorityIdType: GenericAdapter.NOT_APPLICABLE_VALUE,
            managingAuthorityName: GenericAdapter.NOT_APPLICABLE_VALUE,
            nature: ApplicationNature.MONEY,
            object: simplified.intituleProjet.join("|"),
            paymentCondition: null,
            paymentConditionDesc: null,
            paymentId: [simplified.siretDemandeur, ej, simplified.exerciceBudgetaire].join(" - "),
            paymentPeriodDates: null,
            pluriannual: !(simplified.periode.length === 1 && simplified.periode[0] === "PONCTUELLE"),
            pluriannualYears: null,
            provider: "dauphin",
            requestYear: dateDemande?.getFullYear() || null,
            requestedAmount: simplified.montantDemande,
            grantedAmount: simplified.montantAccorde || null,
            scheme: simplified.thematique.join("|") + " - Politique de la ville",
            statusLabel: status,
            subScheme: null,
            subventionPercentage: null,
            totalAmount: null,
            ueNotification: null,
            uniqueId: "dauphin-" + localId + "-" + simplified.exerciceBudgetaire,
            updateDate: new Date(simplified.updateDate),
        };
        return adapted;
    }
}
