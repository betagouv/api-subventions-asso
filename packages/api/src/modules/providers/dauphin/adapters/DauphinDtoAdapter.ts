import { ApplicationStatus, DemandeSubvention } from "@api-subventions-asso/dto";
import DauphinSubventionDto from "../dto/DauphinSubventionDto";
import ProviderValueFactory from "../../../../shared/ProviderValueFactory";
import dauphinService from "../dauphin.service";
import { capitalizeFirstLetter } from "../../../../shared/helpers/StringHelper";
import { toStatusFactory } from "../../helper";
import DauphinGisproDbo from "../repositories/dbo/DauphinGisproDbo";

export default class DauphinDtoAdapter {
    private static _statusConversionArray: { label: ApplicationStatus; providerStatusList: string[] }[] = [
        { label: ApplicationStatus.REFUSED, providerStatusList: ["Rejetée"] },
        {
            label: ApplicationStatus.GRANTED,
            providerStatusList: [
                "A justifier",
                "Justifiée",
                "A été versé",
                "Justification à modifier",
                "Justification en cours",
            ],
        },
        { label: ApplicationStatus.INELIGIBLE, providerStatusList: ["Cloturée", "Non recevable"] },
        {
            label: ApplicationStatus.PENDING,
            providerStatusList: [
                "Prise en charge",
                "Recevable",
                "Transmise",
                "En attente d'attestation",
                "En attente d'instruction",
                "En cours de saisie",
                "En cours",
            ],
        },
    ];

    public static toDemandeSubvention(dbo: DauphinGisproDbo): DemandeSubvention {
        const dauphinData = dbo.dauphin;
        const lastUpdateDate =
            dauphinData._document?.dateVersion ||
            dauphinData.history.events[dauphinData.history.events.length - 1]?.date ||
            dauphinData.history.begin.date;
        const toPV = ProviderValueFactory.buildProviderValueAdapter(
            dauphinService.provider.name,
            new Date(lastUpdateDate),
        );
        const toStatus = toStatusFactory(DauphinDtoAdapter._statusConversionArray);
        const montantDemande = DauphinDtoAdapter.getMontantDemande(dauphinData);
        const montantAccorde = DauphinDtoAdapter.getMontantAccorde(dauphinData);
        let dispositif = "Politique de la ville";
        if (dauphinData.thematique?.title) dispositif = dauphinData.thematique?.title + " - " + dispositif;

        let serviceInstructeur = dauphinData.financeursPrivilegies?.[0].title || "";
        if (serviceInstructeur.match(/^\d{2}-ETAT-POLITIQUE-VILLE/)) {
            const part = serviceInstructeur.split("-");
            serviceInstructeur = `Politique de la ville (Dept ${part[0]})`;
        } else if (serviceInstructeur.match(/^POLITIQUE-VILLE-\d{2,3}/)) {
            const part = serviceInstructeur.split("-");
            serviceInstructeur = `Politique de la ville (Dept ${part[2]})`;
        } else {
            serviceInstructeur = capitalizeFirstLetter(serviceInstructeur.toLowerCase().replace(/-/g, " "));
        }

        return {
            siret: toPV(dauphinData.demandeur.SIRET.complet),
            service_instructeur: toPV("Dauphin"), // Dont use serviceInstructeur because it does not take into account the multifunders
            dispositif: toPV(dispositif),
            ej: dbo.gispro?.ej ? toPV(dbo.gispro?.ej) : undefined,
            versementKey: dbo.gispro?.ej ? toPV(dbo.gispro?.ej) : undefined,
            statut_label: toPV(toStatus(dauphinData.virtualStatusLabel)),
            status: toPV(dauphinData.virtualStatusLabel),
            annee_demande: toPV(dauphinData.exerciceBudgetaire),
            montants: {
                demande: montantDemande ? toPV(montantDemande) : undefined,
                accorde: montantAccorde ? toPV(montantAccorde) : undefined,
            },
            actions_proposee: [
                {
                    intitule: toPV(dauphinData.intituleProjet),
                    objectifs: toPV(dauphinData.description?.value || dauphinData.virtualStatusLabel),
                },
            ],
        };
    }

    private static getMontantDemande(demande: DauphinSubventionDto) {
        return demande.planFinancement
            .find(pf => pf.current)
            ?.recette?.postes?.map(p =>
                p?.sousPostes?.map(s => s?.lignes?.map(l => (l.dispositifEligible ? l.montant.ht : undefined))),
            )
            .flat(2)
            .filter(a => a)[0];
    }

    private static getMontantAccorde(demande: DauphinSubventionDto) {
        return demande.planFinancement
            .find(pf => pf.current)
            ?.recette?.postes?.map(p => p?.sousPostes?.map(s => s?.lignes?.map(l => l.financement?.montantVote?.ht)))
            .flat(2)
            .filter(a => a)[0];
    }
}
