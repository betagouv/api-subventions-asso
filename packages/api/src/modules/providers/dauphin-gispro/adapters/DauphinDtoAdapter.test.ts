const mockLabel = "NORMALIZED_LABEL";
const mockToStatus = jest.fn(() => mockLabel);

import DauphinDtoAdapter from "./DauphinDtoAdapter";

jest.mock("../../providers.adapter", () => ({
    toStatusFactory: () => mockToStatus,
    __esModule: true, // this property makes it work
}));

describe("DauphinDtoAdapter", () => {
    describe("toDocuments", () => {
        const DAUPHIN_DOC = [
            {
                libelle: { value: "test1" },
                documents: [
                    {
                        id: "/path1",
                        title: "title1",
                        expand: { properties: { "entity:document:date": { value: "2022-02-01" } } },
                    },
                    {
                        id: "/path2",
                        title: "title2",
                        expand: { properties: { "entity:document:date": { value: "2022-02-01" } } },
                    },
                ],
            },
            {
                libelle: { value: "test1" },
                documents: [
                    {
                        id: "/path1",
                        title: "title1",
                        expand: { properties: { "entity:document:date": { value: "2022-02-01" } } },
                    },
                    {
                        id: "/path2",
                        title: "title2",
                        expand: { properties: { "entity:document:date": { value: "2022-02-01" } } },
                    },
                ],
            },
        ];

        // not really unit test but don't know how to do otherwise
        it("return proper result", () => {
            // @ts-expect-error -- mock
            const actual = DauphinDtoAdapter.toDocuments(DAUPHIN_DOC);
            expect(actual).toMatchSnapshot();
        });

        it("ignore docs with errors", () => {
            const actual = DauphinDtoAdapter.toDocuments([
                {
                    libelle: { value: "test1" },
                    documents: [
                        {
                            id: "/path2",
                            title: "title2",
                            // @ts-expect-error -- mock
                            error: "something",
                        },
                    ],
                },
            ]);
            expect(actual).toMatchInlineSnapshot(`[]`);
        });
    });
});
