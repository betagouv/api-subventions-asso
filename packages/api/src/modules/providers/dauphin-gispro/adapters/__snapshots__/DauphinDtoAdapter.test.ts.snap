// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DauphinDtoAdapter toDocuments return proper result 1`] = `
[
  {
    "__meta__": {},
    "nom": {
      "last_update": 2022-02-01T00:00:00.000Z,
      "provider": "Dauphin",
      "type": "string",
      "value": "title1",
    },
    "type": {
      "last_update": 2022-02-01T00:00:00.000Z,
      "provider": "Dauphin",
      "type": "string",
      "value": "test1",
    },
    "url": {
      "last_update": 2022-02-01T00:00:00.000Z,
      "provider": "Dauphin",
      "type": "string",
      "value": "/document/dauphin/?url=https%3A%2F%2Fagent-dauphin.cget.gouv.fr%2Fpath1",
    },
  },
  {
    "__meta__": {},
    "nom": {
      "last_update": 2022-02-01T00:00:00.000Z,
      "provider": "Dauphin",
      "type": "string",
      "value": "title2",
    },
    "type": {
      "last_update": 2022-02-01T00:00:00.000Z,
      "provider": "Dauphin",
      "type": "string",
      "value": "test1",
    },
    "url": {
      "last_update": 2022-02-01T00:00:00.000Z,
      "provider": "Dauphin",
      "type": "string",
      "value": "/document/dauphin/?url=https%3A%2F%2Fagent-dauphin.cget.gouv.fr%2Fpath2",
    },
  },
  {
    "__meta__": {},
    "nom": {
      "last_update": 2022-02-01T00:00:00.000Z,
      "provider": "Dauphin",
      "type": "string",
      "value": "title1",
    },
    "type": {
      "last_update": 2022-02-01T00:00:00.000Z,
      "provider": "Dauphin",
      "type": "string",
      "value": "test1",
    },
    "url": {
      "last_update": 2022-02-01T00:00:00.000Z,
      "provider": "Dauphin",
      "type": "string",
      "value": "/document/dauphin/?url=https%3A%2F%2Fagent-dauphin.cget.gouv.fr%2Fpath1",
    },
  },
  {
    "__meta__": {},
    "nom": {
      "last_update": 2022-02-01T00:00:00.000Z,
      "provider": "Dauphin",
      "type": "string",
      "value": "title2",
    },
    "type": {
      "last_update": 2022-02-01T00:00:00.000Z,
      "provider": "Dauphin",
      "type": "string",
      "value": "test1",
    },
    "url": {
      "last_update": 2022-02-01T00:00:00.000Z,
      "provider": "Dauphin",
      "type": "string",
      "value": "/document/dauphin/?url=https%3A%2F%2Fagent-dauphin.cget.gouv.fr%2Fpath2",
    },
  },
]
`;
